src/main/scala/com.snowplowanalytics.maxmind.iplookups/LruMap.scala

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright (c) 2012-2018 Snowplow Analytics Ltd. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+package com.snowplowanalytics.maxmind.iplookups
+
+import cats.effect.Sync
+import java.{util => ju}
+import java.util.LinkedHashMap
+import scala.collection.JavaConverters._
+import scala.collection.mutable.{Map, MapLike, SynchronizedMap}
+
+// Based on com.twitter.util.LruMap
+// https://github.com/twitter/util/blob/develop/util-collection/src/main/scala/com/twitter/util/LruMap.scala
+
+/**
+ * A wrapper trait for java.util.Map implementations to make them behave as scala Maps.
+ * This is useful if you want to have more specifically-typed wrapped objects instead
+ * of the generic maps returned by JavaConverters
+ */
+trait JMapWrapperLike[A, B, +Repr <: MapLike[A, B, Repr] with Map[A, B]]
+    extends Map[A, B]
+    with MapLike[A, B, Repr] {
+  def underlying: ju.Map[A, B]
+
+  override def size = underlying.size
+
+  override def get(k: A) = underlying.asScala.get(k)
+
+  override def +=(kv: (A, B)): this.type = { underlying.put(kv._1, kv._2); this }
+  override def -=(key: A): this.type     = { underlying remove key; this }
+
+  override def put(k: A, v: B): Option[B] = underlying.asScala.put(k, v)
+
+  override def update(k: A, v: B): Unit = underlying.put(k, v)
+
+  override def remove(k: A): Option[B] = underlying.asScala.remove(k)
+
+  override def clear() = underlying.clear()
+
+  override def empty: Repr = null.asInstanceOf[Repr]
+
+  override def iterator = underlying.asScala.iterator
+}
+
+object LruMap {
```

Comment:
could we have a better API like:

```scala
class LruMap[F[_], K, V] private(...) {
  def put(k: K, v: V)(implicit S: Sync[F]): F[Unit]
  def get(k: K)(implciit S: Sync[F]): F[Option[V]]
}

object LruMap {
  def put[F[_]: Sync, K, V](map: LruMap[F, K, V])(k: K, v: V): F[Unit]
  def get[F[_]: Sync, K, V](map: LruMap[F, K, V])(k: K): F[Option[V]]
}
```

wdyt @three @chuwy ?

Reply:
> Is there a difference between def get(k: K)(implicit S: Sync[F]) ...

Good question. No, later is just a syntactic sugar for former one. It is called [context bounds](https://docs.scala-lang.org/tutorials/FAQ/context-bounds.html)

Ben, yep, I like this API, I also think we can omit all other `Map`py methods, such as `keySet`, `map` etc.

Reply:
- `def get(k: K)(implicit S: Sync[F])` lets you do `S.pure` or `S.delay`
- with  `def get[F[_]: Sync](k: K)` you have to do `Sync[F].pure`